cmd.c

```diff
@@ -384,89 +375,81 @@ void cmd_write(char *s)
- * For external dev, name shall be provided. For storage device offset and size shall be provided. */
-#if 0
-static int cmd_parseDev(phfs_conf_t *dev, const char (*args)[LINESZ + 1], u16 *argsID, u16 cmdArgsc)
+static int cmd_parseDev(handler_t *h, addr_t *offs, size_t *sz, const char (*args)[LINESZ + 1], u16 *argsID, u16 argsc)
 {
-	if (plostd_ishex(args[++(*argsID)]) < 0) {
-		plostd_printf(ATTR_LOADER, "\nOpening device %s ..\n", cmd_common.devs[dev->dn].name);
-		dev->handle = phfs_open(dev->dn, args[(*argsID)], 0);
+	const char *alias = args[(*argsID)++];
+
+	if ((*argsID) >= argsc) {
+		plostd_printf(ATTR_ERROR, "\nWrong number of arguments\n");
+		return ERR_ARG;
+	}
 
-		if (dev->handle.h < 0) {
-			plostd_printf(ATTR_ERROR, "Cannot initialize source: %s!\n", cmd_common.devs[dev->dn].name);
+	/* Open device using alias to file */
+	if (plostd_ishex(args[(*argsID)]) < 0) {
+		*offs = 0;
+		*sz = 0;
+		if (phfs_open(alias, args[(*argsID)++], 0, h) < 0) {
+			plostd_printf(ATTR_ERROR, "\nCannot initialize source: %s\n", alias);
 			return ERR_PHFS_IO;
 		}
 	}
+	/* Open device using direct access to memory */
 	else {
 		/* check whether size is provided */
-		if (((*argsID) + 1) >= cmdArgsc || plostd_ishex(args[(*argsID) + 1]) < 0)
-			return -1;
-
-		dev->dataOffs = plostd_ahtoi(args[(*argsID)]);
+		if (((*argsID) + 1) >= argsc || plostd_ishex(args[(*argsID) + 1]) < 0) {
+			plostd_printf(ATTR_ERROR, "\nWrong number of arguments\n");
+			return ERR_ARG;
+		}
 
-		dev->datasz = plostd_ahtoi(args[++(*argsID)]);
+		*offs = plostd_ahtoi(args[(*argsID)]);
+		*sz = plostd_ahtoi(args[++(*argsID)]);
 
-		plostd_printf(ATTR_LOADER, "\nOpening device %s ..\n", cmd_common.devs[dev->dn].name);
-		dev->handle = phfs_open(dev->dn, NULL, 0);
-		if (dev->handle.h < 0) {
-			plostd_printf(ATTR_ERROR, "Cannot initialize source: %s!\n", cmd_common.devs[dev->dn].name);
+		if (phfs_open(alias, NULL, 0, h) < 0) {
+			plostd_printf(ATTR_ERROR, "\nCannot initialize source: %s\n", alias);
 			return ERR_PHFS_IO;
 		}
 	}
 
-	return 0;
+	return ERR_NONE;
 }
-#endif
 
 
 void cmd_copy(char *s)
 {
-#if 0
-	phfs_conf_t phfses[2];
-
-	u16 cmdArgsc = 0, argsID = 0;
-	char cmdArgs[MAX_CMD_ARGS_NB][LINESZ + 1];
-
-	phfses[0].datasz = phfses[1].datasz = 0;
-	phfses[0].dataOffs = phfses[1].dataOffs = 0;
+	size_t sz[2];
+	addr_t offs[2];
+	handler_t h[2];
 
+	u16 argsc = 0, argsID = 0;
+	char args[MAX_CMD_ARGS_NB][LINESZ + 1];
 
 	/* Parse all comand's arguments */
-	if (cmd_parseArgs(s, cmdArgs, &cmdArgsc) < 0 || cmdArgsc < 4) {
+	if (cmd_parseArgs(s, args, &argsc) < 0 || argsc < 4) {
 		plostd_printf(ATTR_ERROR, "\nWrong arguments!!\n");
 		return;
 	}
 
-	/* Parse source parameters */
-	if (cmd_checkDev(cmdArgs[argsID], &phfses[0].dn) < 0)
-		return;
-
-	if (cmd_parseDev(&phfses[0], cmdArgs, &argsID, cmdArgsc) < 0) {
-		phfs_close(phfses[0].dn, phfses[0].handle);
+	if (cmd_parseDev(&h[0], &offs[0], &sz[0], args, &argsID, argsc) < 0) {
+		phfs_close(h[0]);
```

Comment:
If we failed to parseDev do we need to close the handler? I would assume that it shouldn't have been opened in that case, so there is nothing to close.

Reply:
Yes, you have right. I have added information about failure.